### PR TITLE
Fix generic runtime lookup expansion in Tier0

### LINKED ISSUE #101131: Generic dictionary lookup race condition on expansion due to JIT codegen issue
### Description

The situation is that we are in a method compiled for shared generics, which has several specific conditions.

1.	The dictionary layout has a minimum layout size of 4 slots, but an actual layout size of 8. (The exact number is irrelevant, but this is the particular case we have)
2.	There are multiple threads accessing the dictionary of the same type at roughly the same time.
a.	Thread A is attempting to access slot 4
b.	Thread B is attempting to access slot 5
c.	Thread C is attempting to access slot 4
3.	These threads run in the following interleaving
a.	Thread A attempts to access slot 4, determines the value which should be placed in the slot, and gathers the address of the slot which should be updated
b.	Thread B attempts to access slot 5, determines the value which should be placed in the slot, determines that the accessing the slot requires expanding the generic dictionary, expands it, and copies the current state of the slots into the newly expanded dictionary. Notably, since thread A has not written to slot 4, that slot remains as NULL. However, the expanded dictionary is not yet written into the MethodDesc
c.	Thread A fills in slot 4 in the old dictionary with the new value.
d.	Thread C reads the value in slot 4, and determines that it is not NULL, and so commits to the fast path of generic dictionary lookup.
e.	Thread B writes the newly expanded dictionary into the MethodDesc
f.	Thread C then reads the value in slot 4, and since the expanded dictionary does not have a non-NULL value in slot 4, the rest of execution fails unexpectedly.

The issue is caused by incorrect codegen by the JIT, at least in Tier0 mode.

The failing logic is here

```
/_/src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs @ 575:
00007ffb`b9e1b4e0 488b4d10        mov     rcx,qword ptr [rbp+10h]
00007ffb`b9e1b4e4 488b4910        mov     rcx,qword ptr [rcx+10h]
00007ffb`b9e1b4e8 4883792800      cmp     qword ptr [rcx+28h],0
00007ffb`b9e1b4ed 7415            je      Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x514 (00007ffb`b9e1b504)
00007ffb`b9e1b4ef 488b4d10        mov     rcx,qword ptr [rbp+10h]
00007ffb`b9e1b4f3 488b4910        mov     rcx,qword ptr [rcx+10h]
00007ffb`b9e1b4f7 488b4928        mov     rcx,qword ptr [rcx+28h]
00007ffb`b9e1b4fb 48898d10ffffff  mov     qword ptr [rbp-0F0h],rcx
00007ffb`b9e1b502 eb1a            jmp     Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x52e (00007ffb`b9e1b51e)
00007ffb`b9e1b504 488b4d10        mov     rcx,qword ptr [rbp+10h]
00007ffb`b9e1b508 48bab0cef0b9fb7f0000 mov rdx,7FFBB9F0CEB0h
00007ffb`b9e1b512 e8b9b8665e      call    coreclr!JIT_GenericHandleMethod (00007ffc`18486dd0)
00007ffb`b9e1b517 48898510ffffff  mov     qword ptr [rbp-0F0h],rax
00007ffb`b9e1b51e 488b8d10ffffff  mov     rcx,qword ptr [rbp-0F0h]
```

The problem is that the dictionary returned by 
00007ffb`b9e1b4e4 488b4910        mov     rcx,qword ptr [rcx+10h]
Can be updated by another thread, which can cause the actual dictionary entry to be resolved without having a proper NULL check.

There are 2 possibilities for valid code here.
Either we could generate code that looks like…
```
/_/src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs @ 575:
mov     rcx,qword ptr [rbp+10h]
mov     rcx,qword ptr [rcx+10h]
cmp     qword ptr [rcx+28h],0
je      Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x514 (00007ffb`b9e1b504)
mov     rcx,qword ptr [rcx+28h]
mov     qword ptr [rbp-0F0h],rcx
jmp     Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x52e (00007ffb`b9e1b51e)
mov     rcx,qword ptr [rbp+10h]
mov rdx,7FFBB9F0CEB0h
call    coreclr!JIT_GenericHandleMethod (00007ffc`18486dd0)
mov     qword ptr [rbp-0F0h],rax
mov     rcx,qword ptr [rbp-0F0h]
```
Where we only read the dictionary pointer once.

OR
```
/_/src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs @ 575:
mov     rcx,qword ptr [rbp+10h]
mov     rcx,qword ptr [rcx+10h]
mov     rcx,qword ptr [rcx+28h]
cmp     rcx,0
je      Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x514 (00007ffb`b9e1b504)
mov     qword ptr [rbp-0F0h],rcx
jmp     Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.ImmutableArrayExtensions.ToDictionary[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]](System.Collections.Immutable.ImmutableArray`1<System.__Canon>, System.Func`2<System.__Canon,System.__Canon>, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)+0x52e (00007ffb`b9e1b51e)
mov     rcx,qword ptr [rbp+10h]
mov rdx,7FFBB9F0CEB0h
call    coreclr!JIT_GenericHandleMethod (00007ffc`18486dd0)
mov     qword ptr [rbp-0F0h],rax
mov     rcx,qword ptr [rbp-0F0h]
```
Where we read the dictionary entry once.

### Reproduction Steps

Run code which uses many threads, on very subtly different code paths. This particular failure happens in Roslyn, but requires the machine running the code to be running under a tremendous amount of stress.

### Expected behavior

Generic dictionary should be able to expand without triggering a crash.

### Actual behavior

Under extremely rare conditions it will fail.

### Regression?

No, this failing behavior has been in place since 2020.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/101131

Example:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
static void Foo<T>() => Bar<T>();

[MethodImpl(MethodImplOptions.NoInlining)]
static void Bar<T>() { }
```

Codegen diff for `Foo<_Canon>` in Tier0: https://www.diffchecker.com/nVRxo2tr/

tldr: it was an unfortunate "optimization" for Tier0 to reduce code size - instead of spilling a tree to stack and load from it - we used to do load from the context directly again, hence we did it twice - for nullcheck and for the fast path creating a race condition.

The bug is Tier0 only, e.g. here is the codege for `Foo<_Canon>` in Tier1:

```asm
; Assembly listing for method Prog:Foo[System.__Canon]() (FullOpts)
; FullOpts code
       sub      rsp, 40
       mov      qword ptr [rsp+0x20], rcx
       mov      rdx, qword ptr [rcx+0x38]
       mov      rdx, qword ptr [rdx+0x10]
       test     rdx, rdx
       je       SHORT G_M2381_IG04
       jmp      SHORT G_M2381_IG05
G_M2381_IG04:
       mov      rdx, 0x7FFE1F42D630 
       call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
       mov      rdx, rax
G_M2381_IG05:
       mov      rcx, rdx
       call     [Prog:Bar[System.__Canon]()]
       nop      
       add      rsp, 40
       ret      
```

### CODE REVIEW

**Review by jakobbotsch — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
This was new behavior in 8.0, right? Looks like the code moved and I can't quite follow it back to when it was introduced.

**Comment by EgorBo** [comment]
> This was new behavior in 8.0, right? Looks like the code moved and I can't quite follow it back to when it was introduced.

It was introduced in https://github.com/dotnet/runtime/pull/81635 in .NET 8.0, previously, lookup was expanded during import, afair, without this bug.

**Comment by davidwrighton** [comment]
I'm a bit curious what the `if (sizeCheck)` path does. Also, I don't see diffs for Tier1 in the link you sent, contact me tomorrow to point me at them. It seems like it's probably doing what I want, although the code expansion is unfortunate.

**Comment by EgorBo** [comment]
> I'm a bit curious what the `if (sizeCheck)` path does. 

I don't see any issues in it - it reusing a cached indirect as expected, it's a rare case when VM requires size check (`sets `sizeOffset` field in `CORINFO_RUNTIME_LOOKUP`)

> Also, I don't see diffs for Tier1 in the link you sent, contact me tomorrow to point me at them. It seems like it's probably doing what I want, although the code expansion is unfortunate.

The bug is Tier0 only and Tier1 is fine as is, does this match your impressions? E.g. the asm snippets you posted in https://github.com/dotnet/runtime/issues/101131 are for Tier0 (unoptimized) codegen.

**Comment by davidwrighton** [comment]
@EgorBo your comment referred to Tier1 codegen... which I couldn't find any diffs for. For Tier0 yes, it appears like the changes are as expected. It's a bit unfortunate that this results in a spill to the stack, but such is the problem with Tier0 codegen.

**Review by davidwrighton — APPROVED** [review]

**Comment by davidwrighton** [comment]
Also, yes this should probably be ported back to .NET 8.0. Its a stress bug there, and was reported by an internal Microsoft customer using .NET 8.0

**Comment by EgorBo** [comment]
> @EgorBo your comment referred to Tier1 codegen...

Ah I see, I should've used a better wording - I wanted to show that Tier1 is fine and attached a snippet. While for Tier0 I attached a link to diffchecker - https://www.diffchecker.com/nVRxo2tr/ (left - Main, right - this PR)
